File: color/src/main/java/com/tobiasschuerg/color/models/HEXColor.java
package com.tobiasschuerg.color.models;

import android.graphics.Color;
import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;

/**
 * Hue
 * Saturation
 * Value / Brightness
 * <p>
 * Created by Tobias Schürg on 13.08.2016.
 */

public class HEXColor implements ColorModel<HEXColor> {

    @NonNull
    private final String hex;

    @ColorInt
    private final int color;

    public HEXColor(@NonNull String hex) {
        this.hex = hex;
        this.color = Color.parseColor(hex);
    }

    @NonNull
    @Override
    public HEXColor fromColor(int color) {
        return new HEXColor(String.format("#%06X", (0xFFFFFF & color)));

    }

    @NonNull
    @Override
    public HEXColor from(@NonNull ColorModel other) {
        return fromColor(other.toColor());
    }

    @Override
    public int toColor() {
        return color;
    }

    @NonNull
    @Override
    public HSLColor toHSL() {
        return new HSLColor().fromColor(toColor());
    }

    @NonNull
    @Override
    public HSVColor toHSV() {
        return new HSVColor(0, 0, 0).fromColor(color);
    }
}

File: color/src/main/java/com/tobiasschuerg/color/models/HEXColor.kt
package com.tobiasschuerg.color.models

import androidx.annotation.ColorInt
import androidx.core.graphics.toColorInt

/**
 * Hexadecimal color representation.
 *
 * @param color The Android color integer value
 *
 * Created by Tobias Schürg on 13.08.2016.
 */
data class HEXColor(
    @ColorInt private val color: Int,
) : ColorModel {
    constructor(hex: String) : this(hex.toColorInt())

    @ColorInt
    override fun toColor(): Int = color

    override fun equals(other: Any?): Boolean = colorEquals(other)

    override fun hashCode(): Int = colorHashCode()

    /**
     * Returns the hexadecimal string representation of this color.
     * @param includeAlpha Whether to include alpha channel in the output
     * @return Hexadecimal string (e.g., "#FF2196F3" or "#2196F3")
     */
    fun toHexString(includeAlpha: Boolean = true): String {
        return if (includeAlpha) {
            String.format("#%08X", color)
        } else {
            String.format("#%06X", color and 0xFFFFFF)
        }
    }

    override fun toString(): String = toHexString()
}
